enhancements/storage/csi-driver-install.md

```diff
@@ -0,0 +1,181 @@
+
+## Goals
+
+We would like to approach various goals of this KEP in different phases, because we do not expect to get everything right or done in phase-1.
+
+### Phase-1 Goals
+
+* Provide a way for AWS and GCE CSI driver installation . Users could either optionally install it or CSI driver could be installed
+along with in-tree driver and users could use both.
+* Install CSI provided storageclass along with in-tree StorageClass.
+
+### Phase-2 Goals
+
+* Support installation of GCE, Cinder, vSphere, AzureDisk CSI drivers.
+* Provide a way for users to enable in-tree to CSI migration. It basically replaces whole in-tree volume plugin with a less tested CSI implementation. While it's not a strict requirement, we'd like to have the migration optional for at least one release.
+* Provide a way for users to disable in-tree to CSI migration.
+
+### Phase-3 Goals
+
+* Enable CSI drivers as default drivers and CSI provided storageclass as default storageclass.
+* Optionally allow users to configure CSI driver install.
+
+## Non-Goals
+
+* This KEP does not attempt to design installation of third-party or partner CSI drivers.
+* Design of CSI migration enablement.
+  * A new operator will be needed to enable alpha features on API server, controller-manager and kubelets in the right order and with proper node draining before switching the gates.
+  * Right now it's implementation detail, this enhancement focuses on CSI driver installation. Proper enhancement will be created for CSI migration.
+
+## Proposal
+
+We are currently considering using OLM for installation of CSI driver.
+
+### Installation via OLM
+
+We propose that - we provide each driver mentioned above as a separate operator which could be subscribed and installed via OLM UI. Each driver's operator
+is responsible for its installation and release. The operator is responsible for creating storageclass that the driver provides.
+
+The configuration of CSI driver can be done via OLM UI if required and CSI driver can access cloudprovider credentials from Openshift provided sources.The CR that is responsible for driver configuration can be installed by default by the operator itself.
+
+Installation via OLM however means that, when we want to enable these CSI drivers as default drivers, they must be installed by default in Openshift installs.
+We further propose that - Cluster Storage Operator(https://github.com/openshift/cluster-storage-operator) could create subscriptions for these driver operators when drivers have to be installed by default.
+
+Expected workflow as optional driver (using EBS as an example):
+1. User finds EBS CSI driver in OLM and installs it.
```

Comment:
1. User finds EBS CSI driver *operator* in OLM and installs it.
2. The operator automatically creates a singleton CR for itself and install AWS EBS CSI driver with default settings.
    * User can configure basic operator features as log level or ManagementState in the CR. We expect only few (if any) configuration options for the CSI drivers themselves - they don't have any options in-tree.
    * User cannot create any other CR for the operator (ensured via webhooks).

Reply:
Is it common to have the operator installing its own CR?

Can we have the OLM installing the CRD/CR? Then, the webhook would ensure that things are sane by asserting that:

1. The CR can only be installed in the namespace _openshift-aws-ebs-csi-driver-operator_
2. The CR can only have the name _cluster_

This would ensure that there's only 1 CR with this name in this namespace.

Reply:
OLM can install the CRD but it is not usual for OLM to automatically create a CR.

Reply:
I have included this point. Although CR still can't be automatically installed by OLM. It must be created by the operator.